osf/management/commands/data_storage_usage.py

```diff
@@ -0,0 +1,611 @@
+import datetime
+import logging
+import requests
+import tempfile
+import zipfile
+
+from collections import OrderedDict
+from datetime import date
+from decimal import Decimal
+from django.core.management.base import BaseCommand
+from django.db import connection
+from requests_oauthlib import OAuth2
+
+try:
+    from StringIO import StringIO  # Python2
+except ImportError:
+    from io import StringIO  # Python 3
+
+from website.settings import DS_METRICS_BASE_FOLDER, DS_METRICS_OSF_TOKEN
+
+DEFAULT_API_VERSION = '2.14'
+TEMP_FOLDER = tempfile.mkdtemp(suffix='/')
+VALUES = (
+    'file_version_id',
+    'file_id',
+    'version_id',
+    'target_object',
+    'target_content_type_id',
+    'deleted_on',
+    'size',
+    'region',
+    'target_guid',
+    'target_title',
+    'target_type',
+    'target_is_public',
+    'target_is_fork',
+    'target_root',
+    'target_is_deleted',
+    'target_spam_status',
+    'target_is_supplementary_node'
+)
+LAST_ROW_SQL = """
+    select
+        obfnv.id as fileversion_id
+    from osf_basefilenode_versions as obfnv
+    order by obfnv.id DESC
+    LIMIT 1
+"""
+NODE_LIST_SQL = """
+        select
+           obfnv.id as fileversion_id,
+           obfnv.basefilenode_id,
+           obfnv.fileversion_id,
+           file.target_object_id,
+           file.target_content_type_id,
+           file.deleted_on,
+           version.size,
+           region.name as region,
+           guid._id as guid,
+           node.title as node_title,
+           node.type as node_type,
+           node.is_public,
+           node.is_fork,
+           root_guid._id as root_guid,
+           node.is_deleted as node_deleted,
+           node.spam_status,
+           preprint.id is not null as is_supplementary_node
+        from osf_basefilenode_versions as obfnv
+        left outer join osf_basefilenode file on obfnv.basefilenode_id = file.id
+        left outer join osf_fileversion version on obfnv.fileversion_id = version.id
+        left outer join addons_osfstorage_region region on version.region_id = region.id
+        left outer join osf_abstractnode node on node.id = file.target_object_id
+        left outer join osf_preprint preprint on preprint.node_id = node.id
+        LEFT outer JOIN osf_guid guid ON (node.id = guid.object_id AND guid.content_type_id = %s)
+        left outer join osf_guid root_guid ON (node.id = root_guid.object_id AND root_guid.content_type_id = %s)
+        where file.provider = 'osfstorage' and file.target_content_type_id = %s
+        and obfnv.id >= %s and obfnv.id <= %s
+        order by obfnv.id
+    """
+TOTAL_FILE_SIZE_SUM_SQL = """
+    select
+       'total', sum(size) as deleted_size_sum
+    from osf_basefilenode_versions as obfnv
+    left outer join osf_basefilenode file on obfnv.basefilenode_id = file.id
+    left outer join osf_fileversion version on obfnv.fileversion_id = version.id
+    where file.provider = 'osfstorage'
+      and obfnv.id > %s and obfnv.id <= %s
```

Comment:
Since `id` is the primary key of the `osf_basefilenode_versions` table and indexed, one improvement to consider is to do pagination using `ORDER BY` and `LIMIT`:
```
SELECT * FROM ...
    ORDER BY obfnv.id ASC
    WHERE obfnv.id > %s
    LIMIT 1000 (or whatever the page size).
```
This eliminates the need to calculate the `end`, making the code a bit cleaner, and shouldn't affect performance negatively.

Reply:
It's not a bad way to go under normal circumstances. It would still require grabbing the last item in the page to get the id, but that's not a long query, though the actual calculation of the end value is just an addition, so that's not adding any overhead at all.

The biggest problem is that it would involve some correction for how I do the preprint pages, which currently uses the same id range which (probably) allows the database to use cached information when going through that set of data. My suspicion is that it would cost performance in exchange for not very much extra readability.